### ISSUE TITLE
create general workflow for storing metadata in output figures

### ISSUE THREAD

**Issue by zingale** [yt-project/yt#2952]
matplotlib has a way to store metadata in PNGs (using the Agg backend) and PDFs:

https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.pyplot.savefig.html

Is there interest in having this done automatically for figures created with yt?  My idea is that we standardize on a metadata dict that each frontend can populate and when the output images are produced we pass this to the writer to ensure that the metadata is stored in the image.  I know that there was some experimentation with this in the past, but it would be nice to automate it as much as possible.

For instance for our codes, MAESTROeX and Castro, we'd like to store the git hashes of all the source in the output files.  This information is currently available in the frontend.

**Comment by Xarthisius** [comment]
> Is there interest in having this done automatically for figures created with yt?

Yes! I'd even lay a wager that this functionality was added to Matplotlib specifically with yt in mind :)

**Comment by zingale** [comment]
so my proposal would then be to have a `plot_metadata = {}` dict defined by the frontend master class, and then each frontend can do what it wants with it, adding any meta data.  

And then we figure out where to hook this into the writing.

Maybe people can add their implementation thoughts to this issue?  Once I see enough people are happy with it, I can take a pass at implementing it.

**Comment by munkm** [comment]
I'm totally in support! This would be fantastic.

**Comment by cphyc** [comment]
That would be absolutely awesome! One of the things that may be valuable to store as metadata is:
- some frontend-specific data (incl. path to the dataset, etc.)
- the list of filters/plot modifications applied (if applicable)

**Comment by zingale** [comment]
agreed -- we have the name of the compilers, compiler versions, date of compilation, compiler flags, when it was run, where it was run, all of the runtime parameter values, etc. in our frontends, which I envision storing as metadata.

I think I'd defer the second point to a later update.  It's an awesome idea, and one could even imagine storing all of the yt commands used to make the figure in the figure itself.

**Comment by cphyc** [comment]
Maybe one thing we could store systematically is the name of the script that created the file (is it simply `__file__` along with any `argv` that got passed?)

**Comment by Xarthisius** [comment]
I think we shouldn't go too far with our enthusiasm. Some information is better not collected without explicit user consent. Things like paths, filenames etc. are also potential security concerns, not to mention the numerous times I personally used an expletive to name my python script... Would be a shame to end up with a png published in Nature that has curses encoded in its metadata.
